tests/cypress/tests/unit/formatting_test.ts

```diff
@@ -0,0 +1,153 @@
+                expect(formatter(''), 'Empty string case').to.equal('');
+                expect(formatter('foo'), 'Foo string case').to.equal('foo');
+                expect(formatter(true), 'Empty string case').to.equal(true);
+                expect(formatter(NaN), 'NaN case').to.equal('42');
+                expect(formatter(Infinity), 'Infinity case').to.equal('42');
+                expect(formatter(-Infinity), '-Infinity case').to.equal('42');
+                expect(formatter(null as any), 'null case').to.equal('42');
+                expect(formatter(undefined as any), 'undef case').to.equal('42');
+            });
+        });
+
+        describe('with nully handling / partial locale override', () => {
+            it('formats significant digits and grouping separator', () => {
+                const formatter = getFormatter({
+                    locale: getLocale({
+                        decimal: 'x',
+                        group: 'y',
+                        grouping: [2, 1]
+                    }),
+                    nully: getNully('42.4242'),
+                    specifier: getSpecifier(',.2f')
+                });
+
+                assert.isOk(formatter);
+
+                expect(formatter(0)).to.equal('0x00');
+                expect(formatter(0.13)).to.equal('0x13');
+                expect(formatter(0.131)).to.equal('0x13');
+                expect(formatter(1.23)).to.equal('1x23');
+                expect(formatter(1.299)).to.equal('1x30');
+                expect(formatter(1299)).to.equal('1y2y99x00');
+                expect(formatter(1299431)).to.equal('1y2y99y4y31x00');
+                expect(formatter(''), 'Empty string case').to.equal('');
+                expect(formatter('foo'), 'Foo string case').to.equal('foo');
+                expect(formatter(true), 'Empty string case').to.equal(true);
+                expect(formatter(NaN), 'NaN case').to.equal('42.4242');
+                expect(formatter(Infinity), 'Infinity case').to.equal('42.4242');
+                expect(formatter(-Infinity), '-Infinity case').to.equal('42.4242');
+                expect(formatter(null as any), 'null case').to.equal('42.4242');
+                expect(formatter(undefined as any), 'undef case').to.equal('42.4242');
+            });
+        });
+
+        describe('with nully handling / partial locale override with separate_4digits', () => {
+            it('formats significant digits and grouping separator', () => {
+                const formatter = getFormatter({
+                    locale: getLocale({
+                        grouping: [2, 1],
+                        separate_4digits: false
+                    }),
+                    nully: getNully('42.4242'),
+                    specifier: getSpecifier(',.2f')
+                });
+
+                assert.isOk(formatter);
+
+                expect(formatter(-1299)).to.equal('-1299.00');
+                expect(formatter(-1299431)).to.equal('-1,2,99,4,31.00');
```

Comment:
huh, weird to use `separate_4digits` along with `grouping`, thanks for showing this case! Might extend more naturally as `separate_one_digit_more_than_last_group` 🤔 or even a number `min_digits_to_separate` but I guess absent a real user need we should just leave it as is.

Reply:
Yes, I think it's also weird, also considered something along the lines of `one_more_than` and slowly backed away.. without any real thoughts on how it was to behave, I wanted at least a test that would make sure it didn't change behavior without us knowing.